frontend/src/app/feature/room/room.component.ts

```diff
@@ -39,131 +48,126 @@ export class RoomComponent extends WithWebSocket implements OnInit, OnDestroy {
-      this.sendMessage({ type: WsRoomMessageType.CHAT_MESSAGE, data: this.message });
+      this.notifyAll({ type: WsRoomMessageType.CHAT_MESSAGE, data: this.message });
       this.message = '';
     }
   }
 
-  getPlayers(): Array<Player> {
-    return this.room && Array.from<Player>(this.room.players.values())
-      .sort((a, b) => a.id === this.you.id ? -1 : 0);
-  }
-
-  onInventorySave = (inventory: Inventory) => this.sendMessage({ type: WsRoomMessageType.INVENTORY, data: inventory });
+  onInventorySave = (inventory: Inventory) => this.notifyAll({ type: WsRoomMessageType.INVENTORY, data: inventory });
 
   leaveRoom() {
     if (this.wsConnection) {
-      this.onClose();
+      this.onWsClose(); // hmm, looks like a hack
+    }
+    if (this.room && this.currentPlayer) {
+      this.roomService.removePlayerFromRoom(this.room.id, this.currentPlayer.id).subscribe(); // not sure about this, may not work stable
     }
-
     this.router.navigate(['/']);
   }
 
+  protected onWsOpen() {
+    this.notifyAll({ type: WsRoomMessageType.HI_I_AM_NEW_HERE });
+    this.pushMessageToChat('Connected');
+  }
+
+  protected onWsClose() {
+    this.notifyAll({ type: WsRoomMessageType.DISCONNECT });
+  }
+
   protected onMessage(result) {
     const message = JSON.parse(result.data) as WsMessage;
 
     switch (message.type) {
       case WsRoomMessageType.ROLL:
-        this.room.pushMessage(`${this.room.getPlayerById(message.senderId).name}'s Roll result is ${message.data}`);
+        const text = message.data as string;
+        this.pushMessageToChat(`${this.getPlayerByUser(message.senderId).name}'s Roll result is ${text}`);
         break;
 
       case WsRoomMessageType.HI_I_AM_NEW_HERE:
-        if (message.senderId !== this.userService.userId) {
-          this.room.addPlayer(Player.newPlayer(message.data));
-          this.room.pushMessage(`${message.data.name} joined room`);
-          this.sendMessage({
-            type: WsRoomMessageType.NICE_TO_MEET_YOU,
-            data: this.you,
-            direct: true,
-            to: message.senderId
-          });
-        }
-        break;
-
-      case WsRoomMessageType.NICE_TO_MEET_YOU:
-        if (message.direct && message.to === this.userService.userId) {
-          this.room.addPlayer(Player.newPlayer(message.data));
-          this.room.pushMessage(`${message.data.name} already connected`);
+        if (!this.isMyOwnMessage(message)) {
+          const joinedPlayerId = message.senderId;
+          this.playerService.getPlayer(joinedPlayerId).subscribe(p => {
+              this.addPlayerTab(Player.newPlayer(p));
+              this.pushMessageToChat(`${p.name} joined room`);
+            });
         }
         break;
 
       case WsRoomMessageType.CHAT_MESSAGE:
-        this.room.pushMessage(message.data, this.room.getPlayerById(message.senderId));
+        const sender = this.getPlayerByUser(message.senderId);
+        this.pushMessageToChat(message.data, sender);
```

Comment:
it's not that hard to read, we can have it in one line =)

Reply:
Ok